### PR TITLE
osemgrep support for --scan-unknown-extensions

### PR DESCRIPTION
**Pull request by mjambon**
So far, the `Find_targets.conf` config was used only for generic target filtering that was independent of language detection. It is now also used for language detection. The previously unused option `scan_unknown_extensions` was renamed `always_select_explicit_targets` for clarity and now causes language detection to be bypassed for files that are explicit targets.

The set of explicit targets needed somehow to be passed around. I put it as a new field in `Find_targets.conf`. Alternatively, I could have represented each target as a pair (path, is_explicit) but it seemed heavier.

Other: I added a pytest test that exercises the combination of `-e` with `--lang` and `--scan-unknown-extensions`. For the test to pass with osemgrep, I had to populate the `"message"` field of JSON results with the value of the `-e` pattern as it is done by pysemgrep.

### CODE REVIEW

**Comment by aryx** [comment]
lots of CI failures :(
otherwise looks good

File: cli/tests/e2e/snapshots/test_rule_parser/test_rule_parser_cli_pattern/error.txt [hunk 1]
```diff
@@ -3,7 +3,7 @@
 ┌─────────────┐
 │ Scan Status │
 └─────────────┘
-  Scanning 23 files tracked by git with 1 Code rule:
+  Scanning 24 files tracked by git with 1 Code rule:
```
Comment (aryx) [inline cli/tests/e2e/snapshots/test_rule_parser/test_rule_parser_cli_pattern/error.txt:6]: why we suddently have more files scanned in the snapshot? Did you modify pysemgrep file targeting?

File: src/osemgrep/cli_scan/Scan_CLI.ml [hunk 2]
```diff
@@ -245,15 +230,18 @@ in a git repository.
...
+  let default = default.targeting_conf.always_select_explicit_targets in
   H.negatable_flag
     [ "scan-unknown-extensions" ]
     ~neg_options:[ "skip-unknown-extensions" ]
     ~default
     ~doc:
       (spf
-         {|If true, explicit files will be scanned using the language specified
-in --lang. If --skip-unknown-extensions, these files will not be scanned.
+         {|If true, target files specified directly on the command line
```
Comment (aryx) [inline src/osemgrep/cli_scan/Scan_CLI.ml:236]: Far clearer.

Comment (aryx) [reply src/osemgrep/cli_scan/Scan_CLI.ml:236]: cc @armchairlinguist who I think had some beef with this flag in the past.

File: src/osemgrep/cli_scan/Scan_subcommand.ml [hunk 3]
```diff
@@ -535,9 +534,14 @@ let run_scan_files (_caps : < Cap.stdout >) (conf : Scan_CLI.conf)
             Some (file_match_results_hook conf filtered_rules) )
       | { output_conf; _ } -> (output_conf.output_format, None)
     in
+    (* TODO: What is this wrapping for? It makes things really hard to
```
Comment (aryx) [inline src/osemgrep/cli_scan/Scan_subcommand.ml:606]: true

File: src/osemgrep/cli_scan/Scan_subcommand.ml [hunk 4]
```diff
@@ -535,9 +534,14 @@ let run_scan_files (_caps : < Cap.stdout >) (conf : Scan_CLI.conf)
...
+       follow. *)
     let scan_func = mk_scan_func conf file_match_results_hook errors in
     (* step 3': call the engine! *)
     let exn_and_matches =
+      (* TODO: this long code block should not be here!
```
Comment (aryx) [inline src/osemgrep/cli_scan/Scan_subcommand.ml:611]: true. Maybe @ihji can cleanup a bit this code.

File: src/targeting/Find_targets.ml [hunk 5]
```diff
@@ -36,6 +36,28 @@ module Out = Semgrep_output_v1_t
 
 module Fppath_set = Set.Make (Fppath)
 
+module Explicit_targets = struct
```
Comment (aryx) [inline src/targeting/Find_targets.ml:41]: I'm not a huge fan of nested modules. I'd rather move this code in a separate Explicit_targets.ml file.

**Review by aryx — APPROVED** [review]
Very nice! Just need to fix the valid CI regressions (I think that semgrep-pro is failing
is normal since you change the scan_func_for_osemgrep thing.

File: src/osemgrep/cli_scan/Scan_CLI.ml (hunk 2 above)
Comment (armchairlinguist) [reply src/osemgrep/cli_scan/Scan_CLI.ml:236]: If I did, I don't recall it! Sorry 😕

**Comment by aryx** [comment]
conflicts

File: cli/tests/e2e/snapshots/test_rule_parser/test_rule_parser_cli_pattern/error.txt (hunk 1 above)
Comment (mjambon) [reply cli/tests/e2e/snapshots/test_rule_parser/test_rule_parser_cli_pattern/error.txt:6]: I added a file to `cli/tests/e2e/targets/basic/` which is used as the default scanning root in e2e tests.

File: src/targeting/Find_targets.ml (hunk 5 above)
Comment (mjambon) [reply src/targeting/Find_targets.ml:41]: noted.

**Comment by mjambon** [comment]
https://github.com/semgrep/semgrep-proprietary/pull/1317 is looking good. Merging!

File: TCB/TCB.ml [hunk 6]
```diff
@@ -334,9 +334,13 @@ let ( @ ) = ( @ )
  * See also for In: Scanf/Lexing/Parsing
  *)
 
+(*
+I don't know how to use stdin if it's shadowed -- Martin
```
Comment (aryx) [inline TCB/TCB.ml:338]: That's the point. You have to qualify with UStdlib.stdin so it's explicit you're using
the unsafe part of the stdlib (U means unsafe). I'll add a comment about it in TCB.ml

File: libs/commons/UCmd.ml [hunk 7]
```diff
@@ -9,6 +9,21 @@ open Common
...
+(*****************************************************************************)
+(* Helpers *)
+(*****************************************************************************)
+
+(* Log every external command.
+
+   Let's not log environment variables because they may contain sensitive
+   secrets.
+*)
+let log_command cmd =
```
Comment (aryx) [inline libs/commons/UCmd.ml:21]: nice

File: src/core/Scanning_root.mli [hunk 8]
```diff
@@ -0,0 +1,22 @@
+(*
```
Comment (aryx) [inline src/core/Scanning_root.mli:1]: Can this file be moved in src/targeting/ instead?

File: src/core_cli/Core_CLI.ml [hunk 9]
```diff
@@ -425,7 +425,13 @@ let options caps actions =
       Arg.String (fun s -> pattern_file := Some (Fpath.v s)),
       " <file> use the file content as the pattern" );
     ( "-rules",
-      Arg.String (fun s -> rule_source := Some (Rule_file (Fpath.v s))),
+      Arg.String
+        (fun s ->
+          let path = Fpath.v s in
+          (*
+        Printf.eprintf "-rules:\n%s\n%!" (UFile.read_file path);
```
Comment (aryx) [inline src/core_cli/Core_CLI.ml:432]: remove comment or use Logs

File: src/core_cli/Core_CLI.ml [hunk 10]
```diff
@@ -435,7 +441,13 @@ let options caps actions =
       Arg.String (fun s -> lang := Some (Xlang.of_string s)),
       spf " <str> shortcut for -lang" );
     ( "-targets",
-      Arg.String (fun s -> target_source := Some (Target_file (Fpath.v s))),
+      Arg.String
+        (fun s ->
+          let path = Fpath.v s in
+          (*
```
Comment (aryx) [inline src/core_cli/Core_CLI.ml:447]: same

File: src/core_scan/Core_scan_config.ml [hunk 11]
```diff
@@ -44,7 +44,8 @@ type t = {
   rule_source : rule_source option;
   equivalences_file : Fpath.t option;
   lang : Xlang.t option;
-  roots : Fpath.t list;
+  (* Scanning roots. They are mutually exclusive with target_source! *)
```
Comment (aryx) [inline src/core_scan/Core_scan_config.ml:47]: I don't understand this comment.

File: src/osemgrep/cli_publish/Test_publish_subcommand.ml [hunk 12]
```diff
@@ -150,28 +140,14 @@ let test_publish (caps : < Cap.network ; Cap.stdout >) () =
...
-                     single can be published");
-          *)
+          Exit_code.Check.fatal exit_code;
+
           let exit_code =
             Publish_subcommand.main caps
               [| "semgrep-publish"; "--visibility=public"; !!valid_target |]
           in
-          assert (exit_code =*= Exit_code.fatal);
-
```
Comment (aryx) [inline src/osemgrep/cli_publish/Test_publish_subcommand.ml:167]: why remove those comments? They are basically TODOs I think.

File: src/osemgrep/core/Exit_code.ml [hunk 13]
```diff
@@ -4,57 +4,109 @@
...
+type t = { code : int; description : string }
+
+let create code description = { code; description }
+
+(*
+   This systematically logs where the exit code was generated.
+   Otherwise, it's like looking for a needle in haystack.
+*)
+let with_log ~__LOC__:loc x =
+  Logs.info (fun m ->
```
Comment (aryx) [inline src/osemgrep/core/Exit_code.ml:16]: probably better as a Logs.debug

**Review by aryx — COMMENTED** [review]
Nice. Lots of good stuff. Would be better though in the futur to split in different PRs.
one for Exit_code and the ~__LOC, one for the Base comment, one for rrpath types, etc.

**Comment by aryx** [comment]
I'm reverting this. We have 9 error regressions:
```

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
ERROR tests/e2e/test_exclude_include.py::test_exclude_include_verbose_sorted_1
ERROR tests/e2e/test_exclude_include.py::test_exclude_include_verbose_sorted_2
ERROR tests/e2e/test_output.py::test_debug_experimental_rule - Failed: Snapsh...
ERROR tests/e2e/test_semgrep_core_parse_error.py::test_file_parser__failure__error_messages[settings0]
ERROR tests/e2e/test_semgrep_core_parse_error.py::test_file_parser__failure__error_messages[settings1]
ERROR tests/e2e/test_semgrep_core_parse_error.py::test_rule_parser__failure__error_messages[settings0]
ERROR tests/e2e/test_misc.py::test_max_memory - Failed: Snapshot directory wa...
ERROR tests/e2e/test_test.py::test_parse_errors - Failed: Snapshot directory ...
ERROR tests/e2e/test_misc.py::test_timeout_threshold - Failed: Snapshot direc...
= 1565 passed, 1 skipped, 2 xfailed, 12569 warnings, 9 errors in 226.87s (0:03:46) =
make: *** [Makefile:52: test-for-ci] Error 1
Error: Process completed with exit code 2.`
```

**Comment by aryx** [comment]
Maybe the opportunity as you restore this PR to add it piece by piece, like the core_run vs core_scan rename, 
the Rfpath introduction, etc., so it's more incremental.

**Comment by mjambon** [comment]
The run https://github.com/semgrep/semgrep/actions/runs/8120059773/job/22196856080 failed but I didn't see it on Friday. Also, I haven't been able to reproduce these failures today.

**Pull request merged**